githubauth/app.go

```diff
@@ -0,0 +1,380 @@
+	if request.Repositories == nil {
+		return "", fmt.Errorf("requested repositories cannot be nil, did you mean to use AccessTokenAllRepos to request all repos?")
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return "", fmt.Errorf("error marshalling request data: %w", err)
+	}
+
+	return g.githubAccessToken(ctx, requestJSON)
+}
+
+// AccessTokenAllRepos calls the GitHub API to generate a new access token for
+// this application installation with the requested permissions and all granted
+// repositories.
+func (g *App) AccessTokenAllRepos(ctx context.Context, request *TokenRequestAllRepos) (string, error) {
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return "", fmt.Errorf("error marshalling request data: %w", err)
+	}
+
+	return g.githubAccessToken(ctx, requestJSON)
+}
+
+// githubAccessToken calls the GitHub API to generate a new access token with
+// provided JSON payload bytes.
+func (g *App) githubAccessToken(ctx context.Context, requestJSON []byte) (string, error) {
+	appJWT, err := g.AppToken()
+	if err != nil {
+		return "", fmt.Errorf("error generating app jwt: %w", err)
+	}
+	requestURL := fmt.Sprintf(g.accessTokenURLPattern, g.InstallationID)
+
+	requestReader := bytes.NewReader(requestJSON)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, requestURL, requestReader)
+	if err != nil {
+		return "", fmt.Errorf("error creating http request for GitHub installation information: %w", err)
+	}
+	req.Header.Set("Accept", "application/vnd.github+json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", appJWT))
+
+	res, err := g.httpClient.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("error making http request for GitHub installation access token %w", err)
+	}
+	defer res.Body.Close()
+
+	b, err := io.ReadAll(io.LimitReader(res.Body, 64_000))
+	if err != nil {
+		return "", fmt.Errorf("error reading http response for GitHub installation access token %w", err)
+	}
+
+	if res.StatusCode != http.StatusCreated {
+		return "", fmt.Errorf("failed to retrieve token from GitHub - Status: %s - Body: %s", res.Status, string(b))
+	}
+
+	// GitHub will respond with a 201 when you send a request for an invalid combination,
+	// e.g. 'issues':'write' for an empty repository list. This 201 comes with a response that is not actually JSON.
+	// Attempt to parse the JSON to see if this is a valid token, if it is not
+	// then respond with an error containing the
+	// actual response from GitHub.
+	tokenContent := map[string]any{}
+	if err := json.Unmarshal(b, &tokenContent); err != nil {
+		return "", fmt.Errorf("invalid access token from GitHub - Body: %s", string(b))
+	}
+	return string(b), nil
+}
+
+// Token adheres to the oauth2 TokenSource interface and returns a oauth2 token
+// by creating a JWT token.
+func (g *App) OAuthTokenSource() oauth2.TokenSource {
+	return oauth2TokenSource(func() (*oauth2.Token, error) {
+		jwt, err := g.AppToken()
```

Comment:
nit: can we name this more clearly? e.g. OAuthAppTokenSource. Its not immediately clear that you get the app token here and not the installation access token.

Reply:
Done.